### ISSUE TITLE
CMake confused by simultaneous installation of Python2 and Python3

### ISSUE THREAD

**Issue by tammoippen** [nest/nest-simulator#412]
When using older versions of `cmake` (before version 3.4, e.g. [here](https://github.com/Kitware/CMake/blob/v3.3.2/Modules/FindPythonLibs.cmake)) the script to find the Python libraries is independent from the script to find the Python interpreter. Thus, it is possible that Python 2.7 interpreter is found, but NEST is linked against Python 3 libraries. In this case, loading the `pynestkernel` does not work.

With `cmake` 3.4 and following, [`FindPythonLibs`](https://github.com/Kitware/CMake/blob/v3.4.0/Modules/FindPythonLibs.cmake#L54) uses the Python interpreter as hint and looks for the corresponding libraries / include directory.

If you are using an earlier `cmake`, there is a workaround: specify the correct library and include directory on the `cmake` line:

```
cmake  ..... \
    -DPYTHON_LIBRARY=<path/to/python/lib.so> \
    -DPYTHON_INCLUDE_DIR=<path/to/where/Python.h/is/found> \
    <src>
```

To fix this issue, the NEST build system should either figure out the correct library / include or warn the user in case of different versions / prefixes / ... and hint to the workaround.

**Comment by apeyser** [comment]
@jougs @gtrensch 
What cmake are we supporting? If it's >=3.4, we can close this as won't fix.

**Comment by jougs** [comment]
Can we make this so, then? Ubuntu Xenial Xerus(16.04) has 3.5.1. I'd rather not invest time in supporting old packages that can either be gotten from the distribution or source.

If we agree that we drop support for older versions, we need to adapt [INSTALL](https://github.com/nest/nest-simulator/blob/master/INSTALL), which currently says 
> NEST is installed with cmake (at least v2.8.12)

**Comment by tammoippen** [comment]
Keep in mind, that Juqueen and other compute clusters might have older versions installed. We would require users to install cmake in this case.

**Comment by jougs** [comment]
@tammoippen: true.

There are pre-built statically linked versions on the cmake homepage, so installation is not so complicated. but I'm not sure if we should burden people with that if they just happen to have an older distribution.

On the other side, I seem to be the only one hitting this problem regularly and that certainly does not justify to change code.

**Comment by apeyser** [comment]
@jougs 
I hit it regularly (as in today on BGQ, yesterday on my desktop...). 
"CMake considered harmful" is my catch phrase of the day.

I would really suggest only supporting a very narrow band of CMake to limit the support effort we have to put into it. :+1: on requiring (and documenting) cmake >=3.4 (maybe even cmake == 3.4). I don't want to bother with cmake efforts any more --- it's just fundamentally broken.

**Comment by gtrensch** [comment]
Discussed that issue with @apeyser: There is no user-friendly solution. 

We decided to leave the cmake_minimum_required() version check as is for the moment checking for >=2.8.12. Changing this would effect the Travis build.

We will enhance the installation instructions adding a note on the cmake/Python problem and making a recommendation for the cmake version.

**Comment by gtrensch** [comment]
Even though it won't fix the CMake find_package() issue, I found a workaround to prevent the 
version mix when CMake is localizing the Python interpreter, includes and libraries. There are options for the find_package() function allowing version checks. Extracting the interpreter version and rejecting everything which do not exactly match that version works quite well. I tested that on two Ubuntu systems with different Python setups.

I also introduced a new NEST build option: -Dwith-python3=ON/OFF    [default is OFF]
The logic, briefly explained above, also take care of that.

In any case the following CMake built-in variables might be set to leverage auto-detection:
>>  PYTHON_EXECUTABLE ..... path to the Python interpreter
>>  PYTHON_LIBRARY .............. path to libpython
>>  PYTHON_INCLUDE_DIR ..... two include ...
>>  PYTHON_INCLUDE_DIR2 ... directories
